### ISSUE TITLE
Does `<select multiple>` need a button and popup by default?

### ISSUE THREAD

**Issue by josepharhar** [openui/open-ui#1102]
In my mind, the big complication with doing appearance:base for select multiple is that there's not a clear and easy way to represent multiple selected options in the button.

If we follow what browsers already do by having an in-page listbox for select multiple, it would be really easy to go forward with multi selects.

Do most use cases for multi-select have a button? Is it important? If so, what should be in the button?

**Comment by josepharhar** [comment]
ios has a button which says how many items are selected, not sure how good that is though.

![IMG_9375](https://github.com/user-attachments/assets/a564170b-08db-4093-8a89-fddaa03f29e3)

**Comment by josepharhar** [comment]
[This website](https://element-plus.org/en-US/component/select-v2) also has a select multiple which creates buttons for each selected option inside the button. I think it would be very challenging to try to get something like this standardized. This also looks like something that greg was looking into for combobox?

<img width="337" alt="Screenshot 2024-09-27 at 3 02 20 PM" src="https://github.com/user-attachments/assets/96397df6-d7f8-412e-b6e6-04c39eca075f">

**Comment by lukewarlow** [comment]
The above mentioned website is how we designed (one version of) our multi select.

![Screenshot_20240927-230414.png](https://github.com/user-attachments/assets/eb345797-2528-4b9b-8536-2b8b5064dc43)

Chrome Android also does a
 button that says how many items are selected and then has a dialog that opens. But if one item is selected it displays that in the button.

Firefox android has an awkward Inbetween where it renders as a listbox but when you press it it loads a dialog.

Imo an in-page listbox is a separate component that should be able to be single or multi selectable. One of our other versions of multi select used a dual listbox approach where it would be desirable to embedded in the page.

The default multi select should be a button with a drop down imo. We should also try to standardise something like the iOS/chrome android approach. Or perhaps we can try render the selected options as flex items within a `<selectedoptions>` element, that can be styled? More complicated designs such as a close button or being able to deselect by clicking on the tags will require JavaScript to do and probably fall foul of content model guidelines because nested interactivity?

**Comment by lukewarlow** [comment]
https://shoelace.style/components/select#multiple

![Screenshot_20240928-123219.png](https://github.com/user-attachments/assets/1aed218a-9b2c-4716-a7c4-92f72a3bc6b3)

**Comment by brechtDR** [comment]
> The default multi select should be a button with a drop down imo. We should also try to standardise something like the iOS/chrome android approach.

@lukewarlow I agree that this is the most known pattern when looking at external design systems. What do you mean with standardise iOS/android approach?

Isn't this idea to make it fully customisable and then if desired opt-out via CSS again to show native controls if desired? I think CSS style queries could help with further detection in the future, but unfortunately this is currently only usable on custom properties. I might be reading this wrong :)

**Comment by lukewarlow** [comment]
> ios has a button which says how many items are selected.

I mean have this as the default button contents.

**Comment by scottaohara** [comment]
nope. no resolution.

`select multiple` could be a design / updated allowances for children of options (e.g., they can include images) - but otherwise work similar to the current `select multiple` and allow for the listbox to be rendered as an inline / non-popup element.  arguably people could then use script to identify their chosen options and make their own multi-select popup using button/popover.

or `select multiple` works similarly to the use cases mentioned in this thread, where it's a multi-select popup that's rendered.  a few open questions then being:
-  how to handle the display of the selected options in the collapsed state
- how should multi-selection work in the popup?  (a problem with current multi-select popups is that it's not always clear how to exit them.  Some users expect esc to close and _cancel_, reverting to the previous selected option(s) state (or none if none were selected).  Though, various custom implementations
- does a multi-select as a popup actually reuse / repurpose the multiple attribute?  wouldn't that be weird if multiple did one thing for classic select, but for styled select it behaves differently?

**Comment by lukewarlow** [comment]
I think it's worth a discussion on whether we need an embed Vs popup option more generally than select multiple. Because that might sway what we end up doing here.

**Comment by josepharhar** [comment]
All good points Scott, making it a popup is going to be hard.

> does a multi-select as a popup actually reuse / repurpose the multiple attribute? wouldn't that be weird if multiple did one thing for classic select, but for styled select it behaves differently?

One idea I had in mind for this was to use the size attribute to differentiate popup vs in-page since that already happens for single-select, but most mobile browsers already change `<select multiple>` to use a popup without any other opt in which is different from desktop, so the situation is already a bit messed up.

**Comment by lukewarlow** [comment]
One question I have for this. What's the expected screen reader behaviour for multi select. Specifically if it were a drop-down, what would be expected to be announced?

For mobile I believe it is the "N items selected" label. Is this good? It seems it would be too me but I don't use screen readers. 

The alternative is something like

Option 1, Option 3, Option 6... Option N

This presents the actually selected options but for someone who can't see (I know that's not all SR users) it doesn't seem overly helpful every time they land on the select to get a long list read out.

What would be announced for a listbox multi select?

**Comment by davepermen** [comment]
on 'showing the multiple selected vs showing 'x selected''.. maybe this shouldn't be solved by the component but separately. not sure if there's any "on overflow, use this content instead" approach yet to html/css. i do know i could need it in many other places, too.. (similar to details, summary, but automatically switching to summary if the details don't fit).

would give screenreaders both content and allow them to decide intelligently, too. (but that would then be a full own css or html spec thingy ofc)

**Comment by lukewarlow** [comment]
Gonna add to the agenda because I think it'd be good to make progress on solving this question (as it allows us to fix listboxes too)

**Comment by josepharhar** [comment]
Here's my current best strategy for customizable select multiple with popups:
- Require size=1 as an opt-in for select multiple with a popup to solve the mobile vs desktop problem
  - html issue: https://github.com/whatwg/html/issues/8189
  - Here's a table to explain what I mean: https://docs.google.com/presentation/d/10AQxrFI8GIW5q-CYi4mSE4hFK8P3RDGDLWKOdHZb9_A/edit#slide=id.g340018d7cba_1_86
- Make `<selectedcontent>` for `<select multiple>` clone the contents of all selected `<option>`s in tree order and put their contents next to each other.
  - @argyleink pointed out to me that this could be useful if each option had an icon, you could render all of the icons next to each other for example.
- Don't support putting interactive content inside of the author provided `<button>`, just like we do for single-select.
  - This doesn't support use cases like the ones in libraries screenshotted earlier in this thread. I'm not sure if those are even considered accessible or not. If people really want to build those, perhaps they can build their own popover and use an in-page customizable select inside of it.
- Maybe add "OK" and "Cancel" buttons to the bottom of the picker in the UA shadowroot of the popover
  - In response to feedback here: https://github.com/openui/open-ui/issues/1192#issuecomment-2795360353
  - Not sure if this is a good idea or not.
  - We could expose them with pseudo-elements for styling purposes, which would also give authors the ability to display:none them and implement their own activation behavior by preventDefault()ing certain events.

> (as it allows us to fix listboxes too)

I am working on a prototype of appearance:base-select for in-page/listbox selects here, I'll post somewhere to ask for feedback on it when its ready: https://chromium-review.googlesource.com/c/chromium/src/+/5893011

**Comment by lukewarlow** [comment]
Could we not require

Require size!=1 as an opt-in for select multiple with a listbox to solve the mobile vs desktop problem?

That way it's consistent whether you're trying to select 1 or multiple options?

Feels strange to have to be explicit for one and not the other.

**Comment by lukewarlow** [comment]
> Are you saying that <select multiple> should be popup instead of listbox?

Ideally, yes. Though if we can't do that for compat reasons that's okay.

**Comment by lukewarlow** [comment]
```md
> Make <selectedcontent> for <select multiple> clone the contents of all selected <option>s in tree order and put their contents next to each other.
@argyleink pointed out to me that this could be useful if each option had an icon, you could render all of the icons next to each other for example.

How would I for example make this comma separated?

Do we need a container element for each of these? What if an option has multiple root children?
```

**Comment by lukewarlow** [comment]
What should the select announce as when multi select dropdown box? Maybe that helps answer what `<selectedcontent>` should render as

**Comment by josepharhar** [comment]
Here's what i hear on the existing `<select multiple>` on android with this html:
```html
<select multiple>
  <option>red</option>
  <option>blue</option>
</select>
```

When no items are selected: "multiselectable. none selected. zero selected. listbox."

When red is selected: "multiselectable. one of two selected. red. listbox."

When red and blue are selected: "multiselectable. two of two selected. two selected. listbox."

I think the third sentence is just reading out the text content inside the button.

**Comment by josepharhar** [comment]
> Does this make sense as the default state though?

Yes, I think so

**Comment by lukewarlow** [comment]
I won't be here to discuss next week but I'm supportive of the direction Joey proposes though I do still feel it defaulting to combobox would be nicer. As for the button contents I'm still not sure, provided it's a usable default its fine to me.

**Comment by josepharhar** [comment]
Proposed resolution: Select multiple should support rendering with a button and popup in addition to being rendered as an in-page listbox. Controlling which mode is used should be figured out separately.

I proposed how to control these modes here: https://github.com/whatwg/html/issues/8189#issuecomment-2877242732

**Comment by josepharhar** [comment]
I realize now that my proposed resolution doesn't really answer the question in the OP.

I think that the button should always be one button and not have nested buttons or a text field.

In the case that the author doesn't provide a button, it should say "N items selected", or if there is one, then give the text of that option. This matches how select multiple already works on mobile.

**Comment by josepharhar** [comment]
Proposed resolution: Select multiple should support rendering with a button and popup in addition to being rendered as an in-page listbox. The button should not be a text input and should not have nested buttons or other interactive content, just like a single select. Controlling which mode is used should be figured out separately.
